// @mantine/charts/cjs/ScatterChart/ScatterChart.cjs
'use client';
'use strict';

var React = require('react');
var recharts = require('recharts');
var core = require('@mantine/core');
var ChartLegend = require('../ChartLegend/ChartLegend.cjs');
var ChartTooltip = require('../ChartTooltip/ChartTooltip.cjs');
var gridChart_module = require('../grid-chart.module.css.cjs');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

function getAxis(key, dataKey) {
  return key === dataKey.x ? "x" : "y";
}
const defaultProps = {
  withXAxis: true,
  withYAxis: true,
  withTooltip: true,
  tooltipAnimationDuration: 0,
  tickLine: "y",
  strokeDasharray: "5 5",
  gridAxis: "x"
};
const varsResolver = core.createVarsResolver((theme, { textColor, gridColor }) => ({
  root: {
    "--chart-text-color": textColor ? core.getThemeColor(textColor, theme) : void 0,
    "--chart-grid-color": gridColor ? core.getThemeColor(gridColor, theme) : void 0
  }
}));
const ScatterChart = core.factory((_props, ref) => {
  const props = core.useProps("ScatterChart", defaultProps, _props);
  const {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    referenceLines,
    dir,
    withLegend,
    withTooltip,
    withXAxis,
    withYAxis,
    xAxisProps,
    yAxisProps,
    orientation,
    scatterChartProps,
    legendProps,
    data,
    gridAxis,
    tickLine,
    strokeDasharray,
    gridProps,
    tooltipAnimationDuration,
    tooltipProps,
    children,
    onMouseLeave,
    dataKey,
    textColor,
    gridColor,
    xAxisLabel,
    yAxisLabel,
    unit,
    labels,
    valueFormatter,
    scatterProps,
    ...others
  } = props;
  const getFormatter = (axis) => typeof valueFormatter === "function" ? valueFormatter : valueFormatter?.[axis];
  const xFormatter = getFormatter("x");
  const yFormatter = getFormatter("y");
  const theme = core.useMantineTheme();
  const mappedData = data.map((item) => ({
    ...item,
    data: item.data.map((point) => ({ ...point, name: item.name }))
  }));
  const { resolvedClassNames, resolvedStyles } = core.useResolvedStylesApi({
    classNames,
    styles,
    props
  });
  const getStyles = core.useStyles({
    name: "ScatterChart",
    classes: gridChart_module,
    props,
    className,
    style,
    classNames,
    styles,
    unstyled,
    vars,
    varsResolver
  });
  const withXTickLine = gridAxis !== "none" && (tickLine === "x" || tickLine === "xy");
  const withYTickLine = gridAxis !== "none" && (tickLine === "y" || tickLine === "xy");
  const [highlightedArea, setHighlightedArea] = React.useState(null);
  const shouldHighlight = highlightedArea !== null;
  const handleMouseLeave = (event) => {
    setHighlightedArea(null);
    onMouseLeave?.(event);
  };
  const referenceLinesItems = referenceLines?.map((line, index) => {
    const color = core.getThemeColor(line.color, theme);
    return /* @__PURE__ */ React__default.default.createElement(
      recharts.ReferenceLine,
      {
        key: index,
        stroke: line.color ? color : "var(--chart-grid-color)",
        strokeWidth: 1,
        ...line,
        label: {
          value: line.label,
          fill: line.color ? color : "currentColor",
          fontSize: 12,
          position: line.labelPosition ?? "insideBottomLeft"
        },
        ...getStyles("referenceLine")
      }
    );
  });
  const scatters = mappedData.map((item, index) => {
    const dimmed = shouldHighlight && highlightedArea !== item.name;
    return /* @__PURE__ */ React__default.default.createElement(
      recharts.Scatter,
      {
        data: item.data,
        fill: core.getThemeColor(item.color, theme),
        key: index,
        isAnimationActive: false,
        fillOpacity: dimmed ? 0.1 : 1,
        ...scatterProps
      }
    );
  });
  return /* @__PURE__ */ React__default.default.createElement(
    core.Box,
    {
      ref,
      ...getStyles("root"),
      onMouseLeave: handleMouseLeave,
      dir: dir || "ltr",
      ...others
    },
    /* @__PURE__ */ React__default.default.createElement(recharts.ResponsiveContainer, { ...getStyles("container") }, /* @__PURE__ */ React__default.default.createElement(
      recharts.ScatterChart,
      {
        margin: {
          bottom: xAxisLabel ? 30 : void 0,
          left: yAxisLabel ? 10 : void 0,
          right: yAxisLabel ? 5 : void 0
        },
        ...scatterChartProps
      },
      /* @__PURE__ */ React__default.default.createElement(
        recharts.CartesianGrid,
        {
          strokeDasharray,
          vertical: gridAxis === "y" || gridAxis === "xy",
          horizontal: gridAxis === "x" || gridAxis === "xy",
          ...getStyles("grid"),
          ...gridProps
        }
      ),
      /* @__PURE__ */ React__default.default.createElement(
        recharts.XAxis,
        {
          type: "number",
          hide: !withXAxis,
          dataKey: dataKey.x,
          tick: { transform: "translate(0, 10)", fontSize: 12, fill: "currentColor" },
          stroke: "",
          interval: "preserveStartEnd",
          tickLine: withXTickLine ? { stroke: "currentColor" } : false,
          minTickGap: 5,
          unit: unit?.x,
          tickFormatter: xFormatter,
          ...getStyles("axis"),
          ...xAxisProps
        },
        xAxisLabel && /* @__PURE__ */ React__default.default.createElement(recharts.Label, { position: "insideBottom", offset: -20, fontSize: 12, ...getStyles("axisLabel") }, xAxisLabel),
        xAxisProps?.children
      ),
      /* @__PURE__ */ React__default.default.createElement(
        recharts.YAxis,
        {
          type: "number",
          hide: !withYAxis,
          axisLine: false,
          dataKey: dataKey.y,
          tickLine: withYTickLine ? { stroke: "currentColor" } : false,
          tick: { transform: "translate(-10, 0)", fontSize: 12, fill: "currentColor" },
          allowDecimals: true,
          unit: unit?.y,
          tickFormatter: yFormatter,
          ...getStyles("axis"),
          ...yAxisProps
        },
        yAxisLabel && /* @__PURE__ */ React__default.default.createElement(
          recharts.Label,
          {
            position: "insideLeft",
            angle: -90,
            textAnchor: "middle",
            fontSize: 12,
            offset: -5,
            ...getStyles("axisLabel")
          },
          yAxisLabel
        ),
        yAxisProps?.children
      ),
      withTooltip && /* @__PURE__ */ React__default.default.createElement(
        recharts.Tooltip,
        {
          animationDuration: tooltipAnimationDuration,
          isAnimationActive: tooltipAnimationDuration !== 0,
          position: { y: 0 },
          cursor: {
            stroke: "var(--chart-grid-color)",
            strokeWidth: 1,
            strokeDasharray
          },
          content: ({ label, payload }) => /* @__PURE__ */ React__default.default.createElement(
            ChartTooltip.ChartTooltip,
            {
              type: "scatter",
              label,
              payload: labels ? payload?.map((item) => ({
                ...item,
                name: labels[getAxis(item.name, dataKey)] || item.name,
                value: getFormatter(getAxis(item.name, dataKey))?.(item.value) ?? item.value
              })) : payload?.map((item) => ({
                ...item,
                value: getFormatter(getAxis(item.name, dataKey))?.(item.value) ?? item.value
              })),
              classNames: resolvedClassNames,
              styles: resolvedStyles,
              series: data
            }
          ),
          ...tooltipProps
        }
      ),
      withLegend && /* @__PURE__ */ React__default.default.createElement(
        recharts.Legend,
        {
          verticalAlign: "top",
          content: (payload) => /* @__PURE__ */ React__default.default.createElement(
            ChartLegend.ChartLegend,
            {
              payload: payload.payload?.map((item, index) => ({
                ...item,
                dataKey: data[index].name
              })),
              onHighlight: setHighlightedArea,
              legendPosition: legendProps?.verticalAlign || "top",
              classNames: resolvedClassNames,
              styles: resolvedStyles,
              series: data
            }
          ),
          height: 44,
          ...legendProps
        }
      ),
      referenceLinesItems,
      scatters
    ))
  );
});
ScatterChart.displayName = "@mantine/charts/ScatterChart";
ScatterChart.classes = gridChart_module;

exports.ScatterChart = ScatterChart;
