// @mantine/charts/cjs/LineChart/LineChart.cjs
'use client';
'use strict';

var React = require('react');
var recharts = require('recharts');
var core = require('@mantine/core');
var ChartLegend = require('../ChartLegend/ChartLegend.cjs');
var ChartTooltip = require('../ChartTooltip/ChartTooltip.cjs');
var gridChart_module = require('../grid-chart.module.css.cjs');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

const defaultProps = {
  withXAxis: true,
  withYAxis: true,
  withTooltip: true,
  tooltipAnimationDuration: 0,
  fillOpacity: 1,
  tickLine: "y",
  strokeDasharray: "5 5",
  gridAxis: "x",
  withDots: true,
  connectNulls: true,
  strokeWidth: 2,
  curveType: "monotone"
};
const varsResolver = core.createVarsResolver((theme, { textColor, gridColor }) => ({
  root: {
    "--chart-text-color": textColor ? core.getThemeColor(textColor, theme) : void 0,
    "--chart-grid-color": gridColor ? core.getThemeColor(gridColor, theme) : void 0
  }
}));
const LineChart = core.factory((_props, ref) => {
  const props = core.useProps("LineChart", defaultProps, _props);
  const {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    data,
    withLegend,
    legendProps,
    series,
    onMouseLeave,
    dataKey,
    withTooltip,
    withXAxis,
    withYAxis,
    gridAxis,
    tickLine,
    xAxisProps,
    yAxisProps,
    unit,
    tooltipAnimationDuration,
    strokeDasharray,
    gridProps,
    tooltipProps,
    referenceLines,
    withDots,
    dotProps,
    activeDotProps,
    strokeWidth,
    lineChartProps,
    connectNulls,
    fillOpacity,
    curveType,
    orientation,
    dir,
    valueFormatter,
    children,
    lineProps,
    xAxisLabel,
    yAxisLabel,
    ...others
  } = props;
  const theme = core.useMantineTheme();
  const withXTickLine = gridAxis !== "none" && (tickLine === "x" || tickLine === "xy");
  const withYTickLine = gridAxis !== "none" && (tickLine === "y" || tickLine === "xy");
  const [highlightedArea, setHighlightedArea] = React.useState(null);
  const shouldHighlight = highlightedArea !== null;
  const handleMouseLeave = (event) => {
    setHighlightedArea(null);
    onMouseLeave?.(event);
  };
  const { resolvedClassNames, resolvedStyles } = core.useResolvedStylesApi({
    classNames,
    styles,
    props
  });
  const getStyles = core.useStyles({
    name: "LineChart",
    classes: gridChart_module,
    props,
    className,
    style,
    classNames,
    styles,
    unstyled,
    vars,
    varsResolver
  });
  const lines = series.map((item) => {
    const color = core.getThemeColor(item.color, theme);
    const dimmed = shouldHighlight && highlightedArea !== item.name;
    return /* @__PURE__ */ React__default.default.createElement(
      recharts.Line,
      {
        ...getStyles("line"),
        key: item.name,
        name: item.name,
        dataKey: item.name,
        dot: withDots ? { fillOpacity: dimmed ? 0 : 1, strokeOpacity: dimmed ? 0 : 1, ...dotProps } : false,
        activeDot: withDots ? { fill: color, stroke: color, ...activeDotProps } : false,
        fill: color,
        stroke: color,
        strokeWidth,
        isAnimationActive: false,
        fillOpacity: dimmed ? 0 : fillOpacity,
        strokeOpacity: dimmed ? 0.5 : fillOpacity,
        connectNulls,
        type: curveType,
        strokeDasharray: item.strokeDasharray,
        ...typeof lineProps === "function" ? lineProps(item) : lineProps
      }
    );
  });
  const referenceLinesItems = referenceLines?.map((line, index) => {
    const color = core.getThemeColor(line.color, theme);
    return /* @__PURE__ */ React__default.default.createElement(
      recharts.ReferenceLine,
      {
        key: index,
        stroke: line.color ? color : "var(--chart-grid-color)",
        strokeWidth: 1,
        ...line,
        label: {
          value: line.label,
          fill: line.color ? color : "currentColor",
          fontSize: 12,
          position: line.labelPosition ?? "insideBottomLeft"
        },
        ...getStyles("referenceLine")
      }
    );
  });
  return /* @__PURE__ */ React__default.default.createElement(
    core.Box,
    {
      ref,
      ...getStyles("root"),
      onMouseLeave: handleMouseLeave,
      dir: dir || "ltr",
      ...others
    },
    /* @__PURE__ */ React__default.default.createElement(recharts.ResponsiveContainer, { ...getStyles("container") }, /* @__PURE__ */ React__default.default.createElement(
      recharts.LineChart,
      {
        data,
        layout: orientation,
        margin: {
          bottom: xAxisLabel ? 30 : void 0,
          left: yAxisLabel ? 10 : void 0,
          right: yAxisLabel ? 5 : void 0
        },
        ...lineChartProps
      },
      withLegend && /* @__PURE__ */ React__default.default.createElement(
        recharts.Legend,
        {
          verticalAlign: "top",
          content: (payload) => /* @__PURE__ */ React__default.default.createElement(
            ChartLegend.ChartLegend,
            {
              payload: payload.payload,
              onHighlight: setHighlightedArea,
              legendPosition: legendProps?.verticalAlign || "top",
              classNames: resolvedClassNames,
              styles: resolvedStyles,
              series
            }
          ),
          ...legendProps
        }
      ),
      /* @__PURE__ */ React__default.default.createElement(
        recharts.XAxis,
        {
          hide: !withXAxis,
          ...orientation === "vertical" ? { type: "number" } : { dataKey },
          tick: { transform: "translate(0, 10)", fontSize: 12, fill: "currentColor" },
          stroke: "",
          interval: "preserveStartEnd",
          tickLine: withXTickLine ? { stroke: "currentColor" } : false,
          minTickGap: 5,
          ...getStyles("axis"),
          ...xAxisProps
        },
        xAxisLabel && /* @__PURE__ */ React__default.default.createElement(recharts.Label, { position: "insideBottom", offset: -20, fontSize: 12, ...getStyles("axisLabel") }, xAxisLabel),
        xAxisProps?.children
      ),
      /* @__PURE__ */ React__default.default.createElement(
        recharts.YAxis,
        {
          hide: !withYAxis,
          axisLine: false,
          ...orientation === "vertical" ? { dataKey, type: "category" } : { type: "number" },
          tickLine: withYTickLine ? { stroke: "currentColor" } : false,
          tick: { transform: "translate(-10, 0)", fontSize: 12, fill: "currentColor" },
          allowDecimals: true,
          unit,
          tickFormatter: valueFormatter,
          ...getStyles("axis"),
          ...yAxisProps
        },
        yAxisLabel && /* @__PURE__ */ React__default.default.createElement(
          recharts.Label,
          {
            position: "insideLeft",
            angle: -90,
            textAnchor: "middle",
            fontSize: 12,
            offset: -5,
            ...getStyles("axisLabel")
          },
          yAxisLabel
        ),
        yAxisProps?.children
      ),
      /* @__PURE__ */ React__default.default.createElement(
        recharts.CartesianGrid,
        {
          strokeDasharray,
          vertical: gridAxis === "y" || gridAxis === "xy",
          horizontal: gridAxis === "x" || gridAxis === "xy",
          ...getStyles("grid"),
          ...gridProps
        }
      ),
      withTooltip && /* @__PURE__ */ React__default.default.createElement(
        recharts.Tooltip,
        {
          animationDuration: tooltipAnimationDuration,
          isAnimationActive: tooltipAnimationDuration !== 0,
          position: { y: 0 },
          cursor: {
            stroke: "var(--chart-grid-color)",
            strokeWidth: 1,
            strokeDasharray
          },
          content: ({ label, payload }) => /* @__PURE__ */ React__default.default.createElement(
            ChartTooltip.ChartTooltip,
            {
              label,
              payload,
              unit,
              classNames: resolvedClassNames,
              styles: resolvedStyles,
              series,
              valueFormatter
            }
          ),
          ...tooltipProps
        }
      ),
      lines,
      referenceLinesItems,
      children
    ))
  );
});
LineChart.displayName = "@mantine/charts/LineChart";
LineChart.classes = gridChart_module;

exports.LineChart = LineChart;
